arraykit.c

```diff
@@ -129,6 +129,62 @@ AK_ResolveDTypeIter(PyObject *dtypes)
     return resolved;
 }
 
+// Numpy implementation: https://github.com/numpy/numpy/blob/a14c41264855e44ebd6187d7541b5b8d59bb32cb/numpy/core/src/multiarray/methods.c#L1557
+PyObject*
+AK_ArrayDeepCopy(PyArrayObject *array, PyObject *memo)
+{
+    PyObject *id = PyLong_FromVoidPtr((PyObject*)array);
+    if (!id) {
+        return NULL;
+    }
+    PyObject *found = PyDict_GetItem(memo, id);
+    if (found) { // found will be NULL if not in dict
+        Py_INCREF(found); // got a borrowed ref, increment first
+        Py_DECREF(id);
+        return found;
+    }
```

Comment:
Not required, but I generally prefer `PyDict_GetItemWithError`. On the off-chance that something goes wrong, I think it's better not to swallow the exceptions:

```suggestion
    PyObject *found = PyDict_GetItemWithError(memo, id);
    if (found) { // found will be NULL if not in dict
        Py_INCREF(found); // got a borrowed ref, increment first
        Py_DECREF(id);
        return found;
    }
    else if (PyErr_Occurred) {
        goto error;
    }
```

Reply:
I see: since `found` is telling us if the object is in the dict, the only option we have to check for an error is checking `PyErr_Occurred`.

Reply:
I think I need to call PyErr_Occured(); I assume it is what you intended.

```
    else if (PyErr_Occurred()) {
        goto error;
    }
```

Reply:
Yep, that's what I get for writing code in a web browser!

Reply:
Also, I think you might have missed that I suggested changing `PyDict_GetItem` to `PyDict_GetItemWithError`. The former suppresses errors on failure (so you were using it correctly in the original patch), while the latter (which I like better) propagates errors and you need to use `PyErr_Occurred()` to disambiguate.